### PR TITLE
Combine multiple perturbations as 1 test

### PR DESCRIPTION
**Pull request by Prikshit7766**
#### This will be the structure of the config :
````
harness.configure({
    'tests': {
        'defaults': {'min_pass_rate': 0.65},
        'robustness': {
            'add_ocr_typo': {'min_pass_rate': 0.70},
            'multiple_perturbations': {
                'min_pass_rate': 0.50,
                'perturbations1': [
                    'american_to_british',
                    'uppercase',
                    'add_typo'
                ],
                'perturbations2': [
                    'add_ocr_typo',
                    'titlecase'
                ],
                'perturbations3': [
                    'number_to_word',
                    'add_slangs'
                ]
            }
        }
    }
})
````

image

### CODE REVIEW

**Comment by RakshitKhajuria** [comment]
This is the structure of the config :
````
harness.configure({
    'tests': {
        'defaults': {'min_pass_rate': 0.65},
        'robustness': {
            'add_ocr_typo': {'min_pass_rate': 0.70},
            'multiple_perturbations': {
                'min_pass_rate': 0.50,
                'perturbations1': [
                    'american_to_british',
                    'uppercase',
                    'add_typo'
                ],
                'perturbations2': [
                    'add_ocr_typo',
                    'titlecase'
                ],
                'perturbations3': [
                    'number_to_word',
                    'add_slangs'
                ]
            }
        }
    }
})
```

File: nlptest/transform/robustness.py [hunk 1]
```diff
@@ -960,3 +962,79 @@ def slangify_typo(text):
...
+                        new_sample = SequenceClassificationSample(original=sample.test_case, category="robustness",expected_results=sample.expected_results)
+                        new_list.append(new_sample)
+                    transformed_list = apply_transformation(new_list, perturbations[idx])
+
+            for i, sample in enumerate(transformed_list):
+                sample.original = sample_list[i].original
+                sample.transformations=None
+
+
+        elif isinstance(sample_list[0], str):
```
Comment (JulesBelveze) [inline nlptest/transform/robustness.py:1034]: Stupid question: why aren't we performing on `SequenceClassificationSample` and strings?

File: tests/test_robustness.py [hunk 2]
```diff
@@ -197,4 +201,13 @@ def test_add_slangify_typo(self) -> None:
         for sample in transformed_samples:
             self.assertNotEqual(sample.test_case, sample.original)
 
-            
+    def test_multipleperturbations(self) -> None:
+        """"""
```
Comment (JulesBelveze) [inline tests/test_robustness.py:205]: Would be great to add the same tests which inputs of different type than `SequenceClassificationSample`

File: nlptest/transform/__init__.py [hunk 3]
```diff
@@ -331,22 +334,76 @@ def transform(self) -> List[Sample]:
...
+            if TestFactory.task in ('question-answering', 'summarization') and test_name != 'multiple_perturbations' :
                 _ = [
                     sample.transform(test_func, params.get('parameters', {}))
                     if hasattr(sample, 'transform') else sample
                     for sample in data_handler_copy
                 ]
                 transformed_samples = data_handler_copy
 
+            elif test_name == 'multiple_perturbations' and TestFactory.task in ('question-answering', 'summarization'):
+                transformed_samples = []
```
[the remark this answers is not in the document]
Comment (JulesBelveze) [reply nlptest/transform/__init__.py:346]: Gotcha! 
And why do we differently handle `text-classification` and `["question-answering", "summarization"]`?

[the remark this answers is not in the document]
Comment (JulesBelveze) [reply nlptest/transform/__init__.py:346]: Hmmm yeah got it! Thanks @Prikshit7766

File: tests/test_robustness.py (hunk 2 above)
Comment (Prikshit7766) [reply tests/test_robustness.py:205]: inside the  transform/__init__.py
```            if TestFactory.task in ('question-answering', 'summarization') and test_name != 'multiple_perturbations' :
                _ = [
                    sample.transform(test_func, params.get('parameters', {}))
                    if hasattr(sample, 'transform') else sample
                    for sample in data_handler_copy
                ]
                transformed_samples = data_handler_copy
```
here in 'question-answering' and  'summarization' we are not passing complete data at one go here we are passing data record by  record   to  transform method of  BaseQASample class inside the sample.py file 
```
    def transform(self, func, params:None,perturbations=None,**kwargs):
        
        if perturbations is None:
            sens = [self.original_question, self.original_context]
            self.perturbed_question, self.perturbed_context = func(sens, **params, **kwargs)
            self.category = func.__module__.split('.')[-1]

        else:
            sens = [self.original_question, self.original_context]

            self.perturbed_question, self.perturbed_context = func(sens,perturbations, params,**kwargs)
            self.category = func.__module__.split('.')[-1]
```


so if we want to add test for 'question-answering' and  'summarization'   inside the test_robustness.py
so the format will be
```
class RobustnessTestCase(unittest.TestCase):

    def setUp(self) -> None:
        self.multipleperturbations_qa =  [
            "20 euro note -- Until now there has been only one complete series of euro notes; however a new series, similar to the current one, is being released. The European Central Bank will, in due time, announce when banknotes from the first series lose legal tender status.",
            "is the first series 20 euro note still legal tender"
        ]



    def test_multipleperturbations_qa(self) -> None:
        """"""
        transformations = ["lowercase","add_ocr_typo","titlecase","number_to_word"]
       .
       .
       .
      transformed_samples=MultiplePerturbations.transform(self.multipleperturbations_qa,transformations,config=None)
        self.assertIsInstance(transformed_samples, list)

```

Comment (JulesBelveze) [reply tests/test_robustness.py:205]: I think this would be a nice addition and increase test coverage 😄

**Review by JulesBelveze — APPROVED** [review]
LGTM

**Pull request merged**